Firestore/core/src/firebase/firestore/remote/grpc_connection.h

```diff
@@ -35,10 +35,14 @@
 #include "grpcpp/channel.h"
 #include "grpcpp/client_context.h"
 #include "grpcpp/completion_queue.h"
+#ifdef __clang__
```

Comment:
I apologize for how hurriedly I put this together. In my defense I did this to clean up our warnings the day of the 1.0 branch cut.

A better way to handle this is probably with something like this:

```c++
#if defined(__clang__) || defined(__GNUC__)
#define SUPPRESS_BEGIN_UNIMPLEMENTED() \
  _Pragma("GCC diagnostic push")

#define SUPPRESS_BEGIN(name) \
  _Pragma("GCC diagnostic push") \
  _Pragma(name)

#define SUPPRESS_END() \
  _Pragma("GCC diagnostic pop")

#elif defined(_MSC_VER)
#define SUPPRESS_BEGIN_UNIMPLEMENTED() \
  __pragma(warning(push))

#define SUPPRESS_BEGIN(name) \
  __pragma(warning(push)) \
  __pragma(warning(disable:name)

#define SUPPRESS_END() \
  __pragma(warning(pop))

#endif

#if defined(__clang__) || defined(__GNUC__)
#define SUPPRESS_DEPRECATED_DECLARATIONS_BEGIN() \
  SUPPRESS_BEGIN("GCC diagnostic ignored \"-Wdeprecated-declarations\"")

#elif _MSC_VER
#define SUPPRESS_DEPRECATED_DECLARATIONS_BEGIN() \
  SUPPRESS_BEGIN(4995)

#else
#define SUPPRESS_DEPRECATED_DECLARATIONS_BEGIN() \
  SUPPRESS_BEGIN_UNIMPLEMENTED()
#endif

#ifdef __clang__
#define SUPPRESS_DOCUMENTATION_WARNINGS_BEGIN() \
  SUPPRESS_BEGIN("GCC diagnostic ignored \"-Wdocumentation\"")

#else
#define SUPPRESS_DEPRECATED_DECLARATIONS_BEGIN() \
SUPPRESS_BEGIN_UNIMPLEMENTED()
#endif
```

And then instances of this can be become like this:
```c++
SUPRESS_DOCUMENTATION_WARNINGS_BEGIN()
#include "grpcpp/generic/generic_stub.h"
SUPRESS_END()
```
WDYT?

Reply:
I like it. I've made a few changes:
* I've added trailing underscores to the "private" macros.
* I've grouped the "private" and "public" macros together. Note that this does have the effect of splitting the SUPPRRESS_BEGIN_ from SUPPRESS_END.

More details:
* We could prefix everything with FIRESTORE_ to reduce the likelihood of a name collision, but as discussed, we think this is fairly unlikely (and the names are already kinda long).
* I've violated the style guide somewhat, by splitting this header into its own section; after the system includes but before everything else. This allows use to use the macros to protect our includes. (As it happens, the includes we want to protect would naturally occur after the newly added one, but that's not a guarantee for future instances.)
* I've grep'd our code for -Wdocumentation/-Wdeprecated and adjusted to use the new macros.
* I've not tested on msvc.
* The style.sh script did a number on the formatting.

ptal.